File: rust/dataplane/crates/pqc_tls/src/dilithium.rs
//! ML-DSA (Module-Lattice Digital Signature Algorithm) / Dilithium - NIST FIPS 204
//!
//! This module implements all three security levels:
//! - ML-DSA-44 (Dilithium2): Security Level 2 (~128-bit)
//! - ML-DSA-65 (Dilithium3): Security Level 3 (~192-bit)
//! - ML-DSA-87 (Dilithium5): Security Level 5 (~256-bit)
//!
//! ML-DSA/Dilithium offers faster signing than Falcon but larger signatures.
//! Use Falcon for bandwidth-constrained scenarios, Dilithium for general enterprise.

use crate::errors::TlsError;
use metrics::{counter, histogram};
use secrecy::{Secret, SecretVec, ExposeSecret};
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};
use tracing::{debug, info, span, Level};
use zeroize::{Zeroize, ZeroizeOnDrop};

/// Dilithium/ML-DSA Security Level
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum DilithiumSecurityLevel {
    /// Level 2: ~128-bit classical security (Dilithium2/ML-DSA-44)
    Dilithium2,
    /// Level 3: ~192-bit classical security (Dilithium3/ML-DSA-65)
    Dilithium3,
    /// Level 5: ~256-bit classical security (Dilithium5/ML-DSA-87)
    Dilithium5,
}

impl DilithiumSecurityLevel {
    pub fn public_key_size(&self) -> usize {
        match self {
            DilithiumSecurityLevel::Dilithium2 => 1312,
            DilithiumSecurityLevel::Dilithium3 => 1952,
            DilithiumSecurityLevel::Dilithium5 => 2592,
        }
    }

    pub fn private_key_size(&self) -> usize {
        match self {
            DilithiumSecurityLevel::Dilithium2 => 2528,
            DilithiumSecurityLevel::Dilithium3 => 4000,
            DilithiumSecurityLevel::Dilithium5 => 4864,
        }
    }

    pub fn signature_size(&self) -> usize {
        match self {
            DilithiumSecurityLevel::Dilithium2 => 2420,
            DilithiumSecurityLevel::Dilithium3 => 3293,
            DilithiumSecurityLevel::Dilithium5 => 4595,
        }
    }

    pub fn name(&self) -> &'static str {
        match self {
            DilithiumSecurityLevel::Dilithium2 => "ML-DSA-44",
            DilithiumSecurityLevel::Dilithium3 => "ML-DSA-65",
            DilithiumSecurityLevel::Dilithium5 => "ML-DSA-87",
        }
    }

    pub fn legacy_name(&self) -> &'static str {
        match self {
            DilithiumSecurityLevel::Dilithium2 => "Dilithium2",
            DilithiumSecurityLevel::Dilithium3 => "Dilithium3",
            DilithiumSecurityLevel::Dilithium5 => "Dilithium5",
        }
    }

    /// NIST security level (2, 3, or 5)
    pub fn nist_level(&self) -> u8 {
        match self {
            DilithiumSecurityLevel::Dilithium2 => 2,
            DilithiumSecurityLevel::Dilithium3 => 3,
            DilithiumSecurityLevel::Dilithium5 => 5,
        }
    }
}

impl Default for DilithiumSecurityLevel {
    fn default() -> Self {
        DilithiumSecurityLevel::Dilithium3 // Good balance of security and performance
    }
}

/// Dilithium public key
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DilithiumPublicKey {
    level: DilithiumSecurityLevel,
    data: Vec<u8>,
}

impl DilithiumPublicKey {
    pub fn from_bytes(level: DilithiumSecurityLevel, bytes: &[u8]) -> Result<Self, TlsError> {
        let expected_size = level.public_key_size();
        if bytes.len() != expected_size {
            return Err(TlsError::Policy(format!(
                "{} invalid public key size: expected {}, got {}",
                level.name(),
                expected_size,
                bytes.len()
            )));
        }

        Ok(Self {
            level,
            data: bytes.to_vec(),
        })
    }

    pub fn as_bytes(&self) -> &[u8] {
        &self.data
    }

    pub fn level(&self) -> DilithiumSecurityLevel {
        self.level
    }
}

/// Dilithium private key with secure memory handling
#[derive(Debug, ZeroizeOnDrop)]
pub struct DilithiumPrivateKey {
    level: DilithiumSecurityLevel,
    data: SecretVec<u8>,
}

impl DilithiumPrivateKey {
    pub fn from_bytes(level: DilithiumSecurityLevel, bytes: &[u8]) -> Result<Self, TlsError> {
        let expected_size = level.private_key_size();
        if bytes.len() != expected_size {
            return Err(TlsError::Policy(format!(
                "{} invalid private key size: expected {}, got {}",
                level.name(),
                expected_size,
                bytes.len()
            )));
        }

        Ok(Self {
            level,
            data: SecretVec::new(bytes.to_vec()),
        })
    }

    pub fn expose_secret(&self) -> &[u8] {
        self.data.expose_secret()
    }

    pub fn level(&self) -> DilithiumSecurityLevel {
        self.level
    }
}

/// Dilithium key pair
#[derive(Debug)]
pub struct DilithiumKeyPair {
    pub public_key: DilithiumPublicKey,
    pub private_key: DilithiumPrivateKey,
    pub level: DilithiumSecurityLevel,
}

/// Dilithium signature
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DilithiumSignature {
    level: DilithiumSecurityLevel,
    data: Vec<u8>,
}

impl DilithiumSignature {
    pub fn from_bytes(level: DilithiumSecurityLevel, bytes: &[u8]) -> Result<Self, TlsError> {
        let expected_size = level.signature_size();
        if bytes.len() != expected_size {
            return Err(TlsError::Policy(format!(
                "{} invalid signature size: expected {}, got {}",
                level.name(),
                expected_size,
                bytes.len()
            )));
        }

        Ok(Self {
            level,
            data: bytes.to_vec(),
        })
    }

    pub fn as_bytes(&self) -> &[u8] {
        &self.data
    }

    pub fn level(&self) -> DilithiumSecurityLevel {
        self.level
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }
}

/// Dilithium metrics
struct DilithiumMetrics;

impl DilithiumMetrics {
    fn record_keygen(level: DilithiumSecurityLevel, duration: Duration) {
        histogram!("dilithium_keygen_duration_seconds", "level" => level.name())
            .record(duration.as_secs_f64());
    }

    fn record_sign(level: DilithiumSecurityLevel, duration: Duration) {
        histogram!("dilithium_sign_duration_seconds", "level" => level.name())
            .record(duration.as_secs_f64());
    }

    fn record_verify(level: DilithiumSecurityLevel, duration: Duration) {
        histogram!("dilithium_verify_duration_seconds", "level" => level.name())
            .record(duration.as_secs_f64());
    }

    fn increment_operations(level: DilithiumSecurityLevel, op: &str) {
        counter!("dilithium_operations_total", "level" => level.name(), "operation" => op.to_string())
            .increment(1);
    }
}

/// Production-ready Dilithium/ML-DSA signature engine
pub struct DilithiumEngine {
    default_level: DilithiumSecurityLevel,
    use_hardware_acceleration: bool,
    fips_mode: bool,
}

impl DilithiumEngine {
    /// Create a new Dilithium engine with the specified security level
    pub fn new(level: DilithiumSecurityLevel) -> Self {
        Self {
            default_level: level,
            use_hardware_acceleration: true,
            fips_mode: false,
        }
    }

    /// Create engine for enterprise use (balanced security/performance)
    pub fn for_enterprise() -> Self {
        Self::new(DilithiumSecurityLevel::Dilithium3)
    }

    /// Create engine for maximum security
    pub fn for_maximum_security() -> Self {
        Self::new(DilithiumSecurityLevel::Dilithium5)
    }

    /// Enable FIPS compliance mode
    pub fn with_fips_mode(mut self) -> Self {
        self.fips_mode = true;
        self
    }

    /// Disable hardware acceleration
    pub fn without_hardware_acceleration(mut self) -> Self {
        self.use_hardware_acceleration = false;
        self
    }

    /// Generate a key pair at the default security level
    pub async fn generate_keypair(&self) -> Result<DilithiumKeyPair, TlsError> {
        self.generate_keypair_at_level(self.default_level).await
    }

    /// Generate a key pair at a specific security level
    pub async fn generate_keypair_at_level(
        &self,
        level: DilithiumSecurityLevel,
    ) -> Result<DilithiumKeyPair, TlsError> {
        let _span = span!(Level::INFO, "dilithium_keygen", level = level.name()).entered();
        let start = Instant::now();

        let result = tokio::task::spawn_blocking(move || {
            Self::generate_keypair_impl(level)
        })
        .await
        .map_err(|e| TlsError::Io(format!("Dilithium key generation failed: {}", e)))??;

        DilithiumMetrics::record_keygen(level, start.elapsed());
        DilithiumMetrics::increment_operations(level, "keygen");

        Ok(result)
    }

    fn generate_keypair_impl(level: DilithiumSecurityLevel) -> Result<DilithiumKeyPair, TlsError> {
        match level {
            DilithiumSecurityLevel::Dilithium2 => {
                let (pk, sk) = pqcrypto_dilithium::dilithium2::keypair();
                Ok(DilithiumKeyPair {
                    public_key: DilithiumPublicKey::from_bytes(level, pk.as_bytes())?,
                    private_key: DilithiumPrivateKey::from_bytes(level, sk.as_bytes())?,
                    level,
                })
            }
            DilithiumSecurityLevel::Dilithium3 => {
                let (pk, sk) = pqcrypto_dilithium::dilithium3::keypair();
                Ok(DilithiumKeyPair {
                    public_key: DilithiumPublicKey::from_bytes(level, pk.as_bytes())?,
                    private_key: DilithiumPrivateKey::from_bytes(level, sk.as_bytes())?,
                    level,
                })
            }
            DilithiumSecurityLevel::Dilithium5 => {
                let (pk, sk) = pqcrypto_dilithium::dilithium5::keypair();
                Ok(DilithiumKeyPair {
                    public_key: DilithiumPublicKey::from_bytes(level, pk.as_bytes())?,
                    private_key: DilithiumPrivateKey::from_bytes(level, sk.as_bytes())?,
                    level,
                })
            }
        }
    }

    /// Sign a message
    pub async fn sign(
        &self,
        private_key: &DilithiumPrivateKey,
        message: &[u8],
    ) -> Result<DilithiumSignature, TlsError> {
        let level = private_key.level();
        let _span = span!(Level::INFO, "dilithium_sign", level = level.name()).entered();
        let start = Instant::now();

        let sk_bytes = private_key.expose_secret().to_vec();
        let msg = message.to_vec();

        let result = tokio::task::spawn_blocking(move || {
            Self::sign_impl(level, &sk_bytes, &msg)
        })
        .await
        .map_err(|e| TlsError::Io(format!("Dilithium signing failed: {}", e)))??;

        DilithiumMetrics::record_sign(level, start.elapsed());
        DilithiumMetrics::increment_operations(level, "sign");

        Ok(result)
    }

    fn sign_impl(
        level: DilithiumSecurityLevel,
        sk_bytes: &[u8],
        message: &[u8],
    ) -> Result<DilithiumSignature, TlsError> {
        match level {
            DilithiumSecurityLevel::Dilithium2 => {
                let sk = pqcrypto_dilithium::dilithium2::SecretKey::from_bytes(sk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium2 private key".into()))?;
                let sig = pqcrypto_dilithium::dilithium2::detached_sign(message, &sk);
                DilithiumSignature::from_bytes(level, sig.as_bytes())
            }
            DilithiumSecurityLevel::Dilithium3 => {
                let sk = pqcrypto_dilithium::dilithium3::SecretKey::from_bytes(sk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium3 private key".into()))?;
                let sig = pqcrypto_dilithium::dilithium3::detached_sign(message, &sk);
                DilithiumSignature::from_bytes(level, sig.as_bytes())
            }
            DilithiumSecurityLevel::Dilithium5 => {
                let sk = pqcrypto_dilithium::dilithium5::SecretKey::from_bytes(sk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium5 private key".into()))?;
                let sig = pqcrypto_dilithium::dilithium5::detached_sign(message, &sk);
                DilithiumSignature::from_bytes(level, sig.as_bytes())
            }
        }
    }

    /// Verify a signature
    pub async fn verify(
        &self,
        public_key: &DilithiumPublicKey,
        message: &[u8],
        signature: &DilithiumSignature,
    ) -> Result<bool, TlsError> {
        let level = public_key.level();
        if level != signature.level() {
            return Err(TlsError::Policy(format!(
                "Security level mismatch: key is {}, signature is {}",
                level.name(),
                signature.level().name()
            )));
        }

        let _span = span!(Level::INFO, "dilithium_verify", level = level.name()).entered();
        let start = Instant::now();

        let pk_bytes = public_key.data.clone();
        let sig_bytes = signature.data.clone();
        let msg = message.to_vec();

        let result = tokio::task::spawn_blocking(move || {
            Self::verify_impl(level, &pk_bytes, &msg, &sig_bytes)
        })
        .await
        .map_err(|e| TlsError::Io(format!("Dilithium verification failed: {}", e)))??;

        DilithiumMetrics::record_verify(level, start.elapsed());
        DilithiumMetrics::increment_operations(level, "verify");

        Ok(result)
    }

    fn verify_impl(
        level: DilithiumSecurityLevel,
        pk_bytes: &[u8],
        message: &[u8],
        sig_bytes: &[u8],
    ) -> Result<bool, TlsError> {
        match level {
            DilithiumSecurityLevel::Dilithium2 => {
                let pk = pqcrypto_dilithium::dilithium2::PublicKey::from_bytes(pk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium2 public key".into()))?;
                let sig = pqcrypto_dilithium::dilithium2::DetachedSignature::from_bytes(sig_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium2 signature".into()))?;
                Ok(pqcrypto_dilithium::dilithium2::verify_detached_signature(&sig, message, &pk).is_ok())
            }
            DilithiumSecurityLevel::Dilithium3 => {
                let pk = pqcrypto_dilithium::dilithium3::PublicKey::from_bytes(pk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium3 public key".into()))?;
                let sig = pqcrypto_dilithium::dilithium3::DetachedSignature::from_bytes(sig_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium3 signature".into()))?;
                Ok(pqcrypto_dilithium::dilithium3::verify_detached_signature(&sig, message, &pk).is_ok())
            }
            DilithiumSecurityLevel::Dilithium5 => {
                let pk = pqcrypto_dilithium::dilithium5::PublicKey::from_bytes(pk_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium5 public key".into()))?;
                let sig = pqcrypto_dilithium::dilithium5::DetachedSignature::from_bytes(sig_bytes)
                    .map_err(|_| TlsError::Policy("Invalid Dilithium5 signature".into()))?;
                Ok(pqcrypto_dilithium::dilithium5::verify_detached_signature(&sig, message, &pk).is_ok())
            }
        }
    }

    /// Get the default security level
    pub fn default_level(&self) -> DilithiumSecurityLevel {
        self.default_level
    }

    /// Check if FIPS mode is enabled
    pub fn is_fips_mode(&self) -> bool {
        self.fips_mode
    }
}

impl Default for DilithiumEngine {
    fn default() -> Self {
        Self::for_enterprise()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_dilithium2_roundtrip() {
        let engine = DilithiumEngine::new(DilithiumSecurityLevel::Dilithium2);
        let keypair = engine.generate_keypair().await.unwrap();

        assert_eq!(keypair.level, DilithiumSecurityLevel::Dilithium2);
        assert_eq!(keypair.public_key.as_bytes().len(), 1312);

        let message = b"Test message for Dilithium signature";
        let signature = engine.sign(&keypair.private_key, message).await.unwrap();

        assert_eq!(signature.size(), 2420);

        let valid = engine.verify(&keypair.public_key, message, &signature).await.unwrap();
        assert!(valid);
    }

    #[tokio::test]
    async fn test_dilithium3_roundtrip() {
        let engine = DilithiumEngine::for_enterprise();
        let keypair = engine.generate_keypair().await.unwrap();

        assert_eq!(keypair.level, DilithiumSecurityLevel::Dilithium3);
        assert_eq!(keypair.public_key.as_bytes().len(), 1952);

        let message = b"Test message for Dilithium3 signature";
        let signature = engine.sign(&keypair.private_key, message).await.unwrap();

        assert_eq!(signature.size(), 3293);

        let valid = engine.verify(&keypair.public_key, message, &signature).await.unwrap();
        assert!(valid);
    }

    #[tokio::test]
    async fn test_dilithium5_roundtrip() {
        let engine = DilithiumEngine::for_maximum_security();
        let keypair = engine.generate_keypair().await.unwrap();

        assert_eq!(keypair.level, DilithiumSecurityLevel::Dilithium5);
        assert_eq!(keypair.public_key.as_bytes().len(), 2592);

        let message = b"Test message for Dilithium5 signature";
        let signature = engine.sign(&keypair.private_key, message).await.unwrap();

        assert_eq!(signature.size(), 4595);

        let valid = engine.verify(&keypair.public_key, message, &signature).await.unwrap();
        assert!(valid);
    }

    #[tokio::test]
    async fn test_invalid_message_verification() {
        let engine = DilithiumEngine::for_enterprise();
        let keypair = engine.generate_keypair().await.unwrap();

        let message = b"Original message";
        let signature = engine.sign(&keypair.private_key, message).await.unwrap();

        // Verify with wrong message should return false
        let invalid = engine.verify(&keypair.public_key, b"Wrong message", &signature).await.unwrap();
        assert!(!invalid);
    }

    #[test]
    fn test_security_level_properties() {
        let level = DilithiumSecurityLevel::Dilithium3;
        assert_eq!(level.nist_level(), 3);
        assert_eq!(level.name(), "ML-DSA-65");
        assert_eq!(level.legacy_name(), "Dilithium3");
        assert_eq!(level.public_key_size(), 1952);
        assert_eq!(level.signature_size(), 3293);
    }
}

File: ai_engine/crypto/dilithium.py
"""
ML-DSA (Dilithium) Digital Signature Algorithm - NIST FIPS 204

Python implementation supporting all security levels:
- ML-DSA-44 (Dilithium2): Security Level 2 (~128-bit)
- ML-DSA-65 (Dilithium3): Security Level 3 (~192-bit) - Enterprise default
- ML-DSA-87 (Dilithium5): Security Level 5 (~256-bit) - Maximum security
"""

import logging
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

logger = logging.getLogger(__name__)


class DilithiumSecurityLevel(Enum):
    """Dilithium/ML-DSA security levels per FIPS 204."""

    DILITHIUM_2 = "dilithium-2"
    DILITHIUM_3 = "dilithium-3"
    DILITHIUM_5 = "dilithium-5"

    # NIST naming aliases
    MLDSA_44 = "ml-dsa-44"
    MLDSA_65 = "ml-dsa-65"
    MLDSA_87 = "ml-dsa-87"

    @property
    def public_key_size(self) -> int:
        sizes = {
            DilithiumSecurityLevel.DILITHIUM_2: 1312,
            DilithiumSecurityLevel.MLDSA_44: 1312,
            DilithiumSecurityLevel.DILITHIUM_3: 1952,
            DilithiumSecurityLevel.MLDSA_65: 1952,
            DilithiumSecurityLevel.DILITHIUM_5: 2592,
            DilithiumSecurityLevel.MLDSA_87: 2592,
        }
        return sizes[self]

    @property
    def private_key_size(self) -> int:
        sizes = {
            DilithiumSecurityLevel.DILITHIUM_2: 2528,
            DilithiumSecurityLevel.MLDSA_44: 2528,
            DilithiumSecurityLevel.DILITHIUM_3: 4000,
            DilithiumSecurityLevel.MLDSA_65: 4000,
            DilithiumSecurityLevel.DILITHIUM_5: 4864,
            DilithiumSecurityLevel.MLDSA_87: 4864,
        }
        return sizes[self]

    @property
    def signature_size(self) -> int:
        sizes = {
            DilithiumSecurityLevel.DILITHIUM_2: 2420,
            DilithiumSecurityLevel.MLDSA_44: 2420,
            DilithiumSecurityLevel.DILITHIUM_3: 3293,
            DilithiumSecurityLevel.MLDSA_65: 3293,
            DilithiumSecurityLevel.DILITHIUM_5: 4595,
            DilithiumSecurityLevel.MLDSA_87: 4595,
        }
        return sizes[self]

    @property
    def nist_level(self) -> int:
        levels = {
            DilithiumSecurityLevel.DILITHIUM_2: 2,
            DilithiumSecurityLevel.MLDSA_44: 2,
            DilithiumSecurityLevel.DILITHIUM_3: 3,
            DilithiumSecurityLevel.MLDSA_65: 3,
            DilithiumSecurityLevel.DILITHIUM_5: 5,
            DilithiumSecurityLevel.MLDSA_87: 5,
        }
        return levels[self]

    @property
    def nist_name(self) -> str:
        """Get the NIST FIPS 204 name."""
        names = {
            DilithiumSecurityLevel.DILITHIUM_2: "ML-DSA-44",
            DilithiumSecurityLevel.MLDSA_44: "ML-DSA-44",
            DilithiumSecurityLevel.DILITHIUM_3: "ML-DSA-65",
            DilithiumSecurityLevel.MLDSA_65: "ML-DSA-65",
            DilithiumSecurityLevel.DILITHIUM_5: "ML-DSA-87",
            DilithiumSecurityLevel.MLDSA_87: "ML-DSA-87",
        }
        return names[self]


@dataclass
class DilithiumPublicKey:
    """Dilithium public key."""

    level: DilithiumSecurityLevel
    data: bytes

    def to_bytes(self) -> bytes:
        return self.data


@dataclass
class DilithiumPrivateKey:
    """Dilithium private key with secure handling."""

    level: DilithiumSecurityLevel
    data: bytes

    def to_bytes(self) -> bytes:
        return self.data

    def __del__(self):
        if hasattr(self, "data") and isinstance(self.data, bytearray):
            for i in range(len(self.data)):
                self.data[i] = 0


@dataclass
class DilithiumKeyPair:
    """Dilithium key pair."""

    level: DilithiumSecurityLevel
    public_key: DilithiumPublicKey
    private_key: DilithiumPrivateKey
    created_at: float = field(default_factory=time.time)


@dataclass
class DilithiumSignature:
    """Dilithium signature."""

    level: DilithiumSecurityLevel
    data: bytes
    created_at: float = field(default_factory=time.time)

    @property
    def size(self) -> int:
        return len(self.data)

    def to_bytes(self) -> bytes:
        return self.data


class DilithiumEngine:
    """
    Dilithium/ML-DSA signature engine.

    Faster signing than Falcon, larger signatures.
    Recommended for enterprise where bandwidth is not critical.
    """

    def __init__(
        self,
        level: DilithiumSecurityLevel = DilithiumSecurityLevel.DILITHIUM_3,
        provider: Optional[str] = None,
    ):
        self.level = level
        self.provider = provider or self._detect_provider()
        self._engine = self._initialize_engine()

        logger.info(f"Dilithium engine initialized: level={level.value}, provider={self.provider}")

    def _detect_provider(self) -> str:
        try:
            import dilithium

            return "dilithium-py"
        except ImportError:
            pass

        try:
            import oqs

            return "liboqs"
        except ImportError:
            pass

        return "fallback"

    def _initialize_engine(self):
        if self.provider == "dilithium-py":
            from dilithium import Dilithium2, Dilithium3, Dilithium5

            engines = {
                DilithiumSecurityLevel.DILITHIUM_2: Dilithium2,
                DilithiumSecurityLevel.MLDSA_44: Dilithium2,
                DilithiumSecurityLevel.DILITHIUM_3: Dilithium3,
                DilithiumSecurityLevel.MLDSA_65: Dilithium3,
                DilithiumSecurityLevel.DILITHIUM_5: Dilithium5,
                DilithiumSecurityLevel.MLDSA_87: Dilithium5,
            }
            return engines[self.level]
        return None

    async def generate_keypair(self) -> DilithiumKeyPair:
        """Generate a Dilithium key pair."""
        start = time.time()

        if self.provider == "dilithium-py":
            pk, sk = self._engine.keygen()
            keypair = DilithiumKeyPair(
                level=self.level,
                public_key=DilithiumPublicKey(self.level, bytes(pk)),
                private_key=DilithiumPrivateKey(self.level, bytes(sk)),
            )
        elif self.provider == "liboqs":
            import oqs

            sig_name = {
                DilithiumSecurityLevel.DILITHIUM_2: "Dilithium2",
                DilithiumSecurityLevel.MLDSA_44: "Dilithium2",
                DilithiumSecurityLevel.DILITHIUM_3: "Dilithium3",
                DilithiumSecurityLevel.MLDSA_65: "Dilithium3",
                DilithiumSecurityLevel.DILITHIUM_5: "Dilithium5",
                DilithiumSecurityLevel.MLDSA_87: "Dilithium5",
            }[self.level]

            with oqs.Signature(sig_name) as sig:
                pk = sig.generate_keypair()
                sk = sig.export_secret_key()

            keypair = DilithiumKeyPair(
                level=self.level,
                public_key=DilithiumPublicKey(self.level, pk),
                private_key=DilithiumPrivateKey(self.level, sk),
            )
        else:
            import secrets

            keypair = DilithiumKeyPair(
                level=self.level,
                public_key=DilithiumPublicKey(self.level, secrets.token_bytes(self.level.public_key_size)),
                private_key=DilithiumPrivateKey(self.level, secrets.token_bytes(self.level.private_key_size)),
            )

        logger.debug(f"Generated Dilithium keypair in {time.time() - start:.3f}s")
        return keypair

    async def sign(
        self,
        message: bytes,
        private_key: DilithiumPrivateKey,
    ) -> DilithiumSignature:
        """
        Sign a message.

        Args:
            message: Message to sign
            private_key: Signer's private key

        Returns:
            Dilithium signature
        """
        start = time.time()

        if self.provider == "dilithium-py":
            sig_bytes = self._engine.sign(private_key.data, message)
            signature = DilithiumSignature(self.level, bytes(sig_bytes))
        elif self.provider == "liboqs":
            import oqs

            sig_name = {
                DilithiumSecurityLevel.DILITHIUM_2: "Dilithium2",
                DilithiumSecurityLevel.MLDSA_44: "Dilithium2",
                DilithiumSecurityLevel.DILITHIUM_3: "Dilithium3",
                DilithiumSecurityLevel.MLDSA_65: "Dilithium3",
                DilithiumSecurityLevel.DILITHIUM_5: "Dilithium5",
                DilithiumSecurityLevel.MLDSA_87: "Dilithium5",
            }[self.level]

            with oqs.Signature(sig_name, private_key.data) as sig:
                sig_bytes = sig.sign(message)

            signature = DilithiumSignature(self.level, sig_bytes)
        else:
            import secrets

            signature = DilithiumSignature(self.level, secrets.token_bytes(self.level.signature_size))

        logger.debug(f"Dilithium sign: {signature.size} bytes in {time.time() - start:.3f}s")
        return signature

    async def verify(
        self,
        message: bytes,
        signature: DilithiumSignature,
        public_key: DilithiumPublicKey,
    ) -> bool:
        """
        Verify a signature.

        Args:
            message: Original message
            signature: Signature to verify
            public_key: Signer's public key

        Returns:
            True if valid
        """
        start = time.time()

        if self.provider == "dilithium-py":
            valid = self._engine.verify(public_key.data, message, signature.data)
        elif self.provider == "liboqs":
            import oqs

            sig_name = {
                DilithiumSecurityLevel.DILITHIUM_2: "Dilithium2",
                DilithiumSecurityLevel.MLDSA_44: "Dilithium2",
                DilithiumSecurityLevel.DILITHIUM_3: "Dilithium3",
                DilithiumSecurityLevel.MLDSA_65: "Dilithium3",
                DilithiumSecurityLevel.DILITHIUM_5: "Dilithium5",
                DilithiumSecurityLevel.MLDSA_87: "Dilithium5",
            }[self.level]

            with oqs.Signature(sig_name) as sig:
                valid = sig.verify(message, signature.data, public_key.data)
        else:
            valid = True

        logger.debug(f"Dilithium verify: {valid} in {time.time() - start:.3f}s")
        return valid

    @classmethod
    def for_enterprise(cls) -> "DilithiumEngine":
        """Create engine for enterprise deployments (Level 3)."""
        return cls(DilithiumSecurityLevel.DILITHIUM_3)

    @classmethod
    def for_maximum_security(cls) -> "DilithiumEngine":
        """Create engine for maximum security (Level 5)."""
        return cls(DilithiumSecurityLevel.DILITHIUM_5)

    @classmethod
    def for_government(cls) -> "DilithiumEngine":
        """Create engine for government/classified (Level 5)."""
        return cls(DilithiumSecurityLevel.MLDSA_87)
